src/serac/physics/utilities/variational_form/integral.hpp

```diff
@@ -0,0 +1,597 @@
+ * TODO: L2 family elements integrate f against test space shape functions
+ *
+ * In the future, the user will be able to override these defaults
+ * to omit unused components (e.g. provide only the term to be integrated against test function gradients)
+ * @tparam element_type The type of the element (used to determine the family)
+ * @tparam T The type of the output from the user-provided q-function
+ * @pre T must be a pair type for H1, H(curl) and H(div) family elements
+ * @param[in] f The value component output of the user's quadrature function (as opposed to the value/derivative pair)
+ * @param[in] xi The position of the quadrature point in reference space
+ * @param[in] J The Jacobian of the element transformation at the quadrature point
+ */
+template <typename element_type, typename T, int dim>
+auto Postprocess(T f, const tensor<double, dim> xi, const tensor<double, dim, dim> J)
+{
+  // TODO: Helpful static_assert about f being tuple or tuple-like for H1, hcurl, hdiv
+  if constexpr (element_type::family == Family::H1) {
+    auto W     = element_type::shape_functions(xi);
+    auto dW_dx = dot(element_type::shape_function_gradients(xi), inv(J));
+    return outer(W, std::get<0>(f)) + dot(dW_dx, std::get<1>(f));
+  }
+
+  if constexpr (element_type::family == Family::HCURL) {
+    auto W      = dot(element_type::shape_functions(xi), inv(J));
+    auto curl_W = element_type::shape_function_curl(xi) / det(J);
+    if constexpr (dim == 3) {
+      curl_W = dot(curl_W, transpose(J));
+    }
+    return (W * std::get<0>(f) + curl_W * std::get<1>(f));
+  }
+}
+
+/**
+ * @overload
+ * @note This specialization of impl::Postprocess is called when doing integrals
+ * where the spatial dimension is different from the dimension of the element geometry
+ * (i.e. surface integrals in 3D space, line integrals in 2D space, etc)
+ *
+ * In this case, q-function outputs are only integrated against test space shape functions
+ * QUESTION: Should test function gradients be supported here or not?
+ */
+template <typename element_type, typename T, int geometry_dim, int spatial_dim>
+auto Postprocess(T f, const tensor<double, geometry_dim> xi,
+                 [[maybe_unused]] const tensor<double, spatial_dim, geometry_dim> J)
+{
+  if constexpr (element_type::family == Family::H1) {
+    return outer(element_type::shape_functions(xi), f);
+  }
+
+  if constexpr (element_type::family == Family::HCURL) {
+    return outer(element_type::shape_functions(xi), dot(inv(J), f));
+  }
+}
+
+/**
+ * @brief Takes in a Jacobian matrix and computes the
+ * associated length / area / volume ratio of the transformation
+ *
+ * In general, this is given by sqrt(det(J^T * J)), but for the case
+ * where J is square, this is equivalent to just det(J)
+ *
+ * @param[in] A The Jacobian to compute the ratio on
+ */
+
+template <int n>
+auto Measure(const tensor<double, n, n>& A)
```

Comment:
I think renaming this to `Determinant` or `Det` would be more explicit as there can be many measures.

Reply:
I'm happy to rename things to improve clarity, but `Measure` needs to work for rectangular matrices too (which should rule out `determinant`).

`Measure` is just what Mathematica uses to refer to the length / area / volume of a geometric region, but if there is a better term, let's use that instead.

Reply:
Hmm, a fair point. As this is what mathematica uses, I retract my statement.

Reply:
Some doxygen comments as to what `Measure` is acutally doing for the various dimensions is needed however.

Reply:
`Measure` is already be doxygen commented, but I'm not sure what the right convention is for doing doxygen comments on overload sets. All the functions in an overload set do the same thing conceptually, with slight differences specific to the argument types. Should doxygen comments for overloaded functions:
1. repeat the conceptual preamble which is common to all of the overloads
2. focus on just the part that is specific to that overload
3. do nothing and just have a large comment on the first function in the overload set that explains the differences

Reply:
My preference would be something like the following (this is what I tried to do in some of the other functions for this branch):

```cpp

/**
 * @brief What all the functions do/have in common
 * Things specific for this @p int function
 */
void overloaded(int);

/**
 * @overload
 * Things specific for this @p float function
 */
void overloaded(float);

// ... et cetera

```

Doxygen will render the subsequent functions (those with the `overload` command) with the message "this is an overload.  it differs from the above function in the arguments that it accepts" or something to that effect

Reply:
This function computes a [pushforward measure](src/serac/physics/utilities/variational_form/integral.hpp), so I'd favor naming it `PushforwardMeasure` instead of just `Measure`, but that's partly because I could see an outside chance of extending this functionality to incorporate non-Lebesgue measures, which come up often in UQ.

Reply:
I think this is a bit in-the-weeds for your average user. My preference is to keep it as is for now, and possibly rename it if the need arises.